consensus/polybft/consensus_runtime.go

```diff
@@ -0,0 +1,851 @@
+type consensusRuntime struct {
+	// config represents wrapper around required parameters which are received from the outside
+	config *runtimeConfig
+
+	// state is reference to the struct which encapsulates bridge events persistence logic
+	state *State
+
+	// eventTracker is a reference to the log event tracker
+	eventTracker *eventTracker
+
+	// epoch is the metadata for the current epoch
+	epoch *epochMetadata
+
+	// lock is a lock to access 'epoch'
+	lock sync.Mutex
+
+	// lastBuiltBlock is the header of the last processed block
+	lastBuiltBlock *types.Header
+
+	// activeValidatorFlag indicates whether the given node is amongst currently active validator set
+	activeValidatorFlag uint32
+
+	logger hcf.Logger
+}
+
+// newConsensusRuntime creates and starts a new consensus runtime instance with event tracking
+func newConsensusRuntime(log hcf.Logger, config *runtimeConfig) (*consensusRuntime, error) {
+	runtime := &consensusRuntime{
+		state:  config.State,
+		config: config,
+		logger: log.Named("consensus_runtime"),
+	}
+
+	if runtime.IsBridgeEnabled() {
+		err := runtime.startEventTracker()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return runtime, nil
+}
+
+// getEpoch returns current epochMetadata in a thread-safe manner.
+func (c *consensusRuntime) getEpoch() *epochMetadata {
+	c.lock.Lock()
+	epoch := c.epoch
+	c.lock.Unlock()
+
+	return epoch
+}
+
+func (c *consensusRuntime) IsBridgeEnabled() bool {
+	return c.config.PolyBFTConfig.IsBridgeEnabled()
+}
+
+// AddLog is an implementation of eventSubscription interface,
+// and is called from the event tracker when an event is final on the rootchain
+func (c *consensusRuntime) AddLog(eventLog *ethgo.Log) {
+	c.logger.Info(
+		"Add State sync event",
+		"block", eventLog.BlockNumber,
+		"hash", eventLog.TransactionHash,
+		"index", eventLog.LogIndex,
+	)
+
+	event, err := decodeEvent(eventLog)
+	if err != nil {
+		c.logger.Error("failed to decode state sync event", "hash", eventLog.TransactionHash, "err", err)
+
+		return
+	}
+
+	if err := c.state.insertStateSyncEvent(event); err != nil {
+		c.logger.Error("failed to insert state sync event", "hash", eventLog.TransactionHash, "err", err)
+
+		return
+	}
+	// TODO: Nemanja
+	// update metrics
+	// totalStateSyncsMeter.Mark(1)
+	return // TODO: Delete this when metrics is established. This is added just to trick linter.
+}
+
+// NotifyProposalInserted is an implementation of fsmNotify interface
+func (c *consensusRuntime) NotifyProposalInserted(b *StateBlock) {
+	lastHeader := b.Block.Header
+	if c.isEndOfEpoch(lastHeader.Number) {
+		// reset the epoch. Internally it updates the parent block header.
+		if err := c.restartEpoch(lastHeader); err != nil {
+			c.logger.Error("failed to restart epoch after block inserted", "err", err)
+		}
+	} else {
+		// inside the epoch, update last built block header
+		c.lastBuiltBlock = lastHeader
+	}
```

Comment:
```go
	lastHeader := b.Block.Header
	if !c.isEndOfEpoch(lastHeader.Number) {
		// Inside the epoch, update last built block header
		c.lastBuiltBlock = lastHeader

		return
	}

	// Reset the epoch. Internally it updates the parent block header
	if err := c.restartEpoch(lastHeader); err != nil {
		c.logger.Error("failed to restart epoch after block inserted", "err", err)
	}

```

Reply:
this is just coding style approach. else usage is totally fine with me and more readable, for example